example/bwmf/bwmf_task.go

```diff
@@ -66,133 +50,398 @@ type bwmfTask struct {
+	// fullT is ready here
 
-// Read dShard and tShard from last checkpoint if any.
-func (t *bwmfTask) readLastCheckpoint() bool {
-	panic("")
-}
+	// if dLoss is not initialized
+	if bt.dLoss.n == -1 {
+		bt.dLoss.n = 0
+		for _, m := range bt.fullT {
+			bt.dLoss.n += len(m.Row)
+		}
+		bt.N = bt.dLoss.n
+		wh := make([][]float32, bt.N)
+		for i, _ := range wh {
+			wh[i] = make([]float32, bt.m)
+		}
+		bt.dLoss.WH = wh
+	}
+	bt.dLoss.W = NewBlocksParameter(&bt.fullT)
 
-// Task have all the data, compute local optimization of D/T.
-func (t *bwmfTask) localCompute() {}
-
-// This is useful to bring the task up to speed from scratch or if it recovers.
-func (t *bwmfTask) Init(taskID uint64, framework taskgraph.Framework) {
-	t.taskID = taskID
-	t.framework = framework
-	t.logger = log.New(os.Stdout, "", log.Ldate|log.Ltime|log.Lshortfile)
-	// Use some unique identifier to set Index "in the future".
-	// We can use taskID now.
-	t.dShard = newDShard(t.taskID)
-	t.tShard = newTShard(t.taskID)
-	t.readShardsFromDisk()
-	ok := t.readLastCheckpoint()
+	// all set. Start the optimizing routine.
+	ok := bt.optimizer.Minimize(bt.dLoss, bt.stopCriteria, bt.shardedDParam)
 	if !ok {
-		t.dShard.randomFillValue()
-		t.tShard.randomFillValue()
+		// TODO report error
 	}
 
-	// At initialization:
-	// Task 0 will start the iterations.
+	// signal that dShard has already been evaluated
+	bt.dDone <- &event{epoch: bt.epoch}
+}
+
+func (bt *bwmfTask) updateTShard() {
+	// Symmetric as updateDShard
+	if bt.tLoss.m == -1 {
+		bt.tLoss.m = 0
+		for _, m := range bt.fullD {
+			bt.tLoss.m += len(m.Row)
+		}
+		bt.M = bt.tLoss.m
+		wh := make([][]float32, bt.M)
+		for i, _ := range wh {
+			wh[i] = make([]float32, bt.n)
+		}
+		bt.tLoss.WH = wh
+	}
+	bt.tLoss.W = NewBlocksParameter(&bt.fullD)
+	ok := bt.optimizer.Minimize(bt.tLoss, bt.stopCriteria, bt.shardedTParam)
+	if !ok {
+		// TODO report error
+	}
+	bt.tDone <- &event{epoch: bt.epoch}
 }
 
-func (t *bwmfTask) Exit() {}
+func (bt *bwmfTask) Init(taskID uint64, framework taskgraph.Framework) {
+	bt.taskID = taskID
+	bt.framework = framework
+	bt.logger = log.New(os.Stdout, "", log.Ldate|log.Ltime|log.Lshortfile)
 
-func (t *bwmfTask) SetEpoch(ctx taskgraph.Context, epoch uint64) {
-	t.logger.Printf("slave SetEpoch, task: %d, epoch: %d\n", t.taskID, epoch)
-	t.epoch = epoch
-	t.childrenReady = make(map[uint64]bool)
-	t.dtReady = common.NewCountdownLatch(int(t.numOfTasks))
+	// Below is loading the file. NOTE: this should be moved to a separate function.
+	hdfsClient, hcOk := filesystem.NewHdfsClient(bt.namenodeAddr, bt.webHdfsAddr, bt.hdfsUser)
+	if hcOk != nil {
+		bt.logger.Fatal("Failed connecting to HDFS at ", bt.namenodeAddr, bt.webHdfsAddr, ", with account ", bt.hdfsUser)
+	}
+	rowShardReader, rsrOk := hdfsClient.OpenReadCloser(bt.rowShardPath)
+	if rsrOk != nil {
+		bt.logger.Fatalf("Failed load matrix data on task: %d, form path %s", bt.taskID, bt.rowShardPath)
```

Comment:
"task: %d"
-> "task %d"

I make all logging like this. This will help debugging.